Decentralized pallet-staking audit tool

taqtiqa-mark opened the pull request:
#### Project Abstract

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [x] **Level 3**:  Unlimited, 5 approvals

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

Noc2 reviewed (changes requested):
Thanks a lot for the application, and I'm sorry for the delay here. I will mark it as ready for review and share it with the rest of the team. But I have one initial question: Why was the price suddenly increased from 16k to 75k?  I also find the estimated workload to develop an Excel/OpenOffice workbook, as well as the daily rate, very high to be honest.

laboon commented on applications/decentralized-pallet-staking-audit-tool.md:
This is incorrect. Staking returns, as mentioned on that page, are based on the ideal staking rate. That chart is based on an ideal staking rate of 50%. The actual ideal staking rate is 45% on Polkadot right now (note that there is a bug on PJS UI showing it at 75%). With a current staking rate of ~ 49%, we're on the steep slope to the right, and just eyeballing it, a bit less than 15% seems correct.

taqtiqa-mark replied:
> This is incorrect.

That is my working assumption :)
Lets see...

> With a current staking rate of ~ 49%, we're on the steep slope to the right, and just eyeballing it, a bit less than 15% seems correct.

That chart is not labeled, which does not help. The horizontal axis is either

a) the staking rate: My calculation/expectation
b) or 1-the staking rate: Your calculation/expectation

Wouldn't it be useful if there was a tool we both had that gave the same answer :)

Anyway, back to the problem. From the text immediately above the chart:

> If the amount of tokens staked goes below the ideal rate, then staking rewards for nominators go up...

In your telling Polkadot staking is below the ideal rate and so we should expect a reward % above 20%. Why?
Again, from the text above the chart:

> ...incentivizing them to stake more tokens on the network. On the contrary, if it goes above the ideal rate, staking rewards drop.

Hopefully I haven't got that back to front.

Agreed?

taqtiqa-mark replied:
> This is incorrect. Staking returns, as mentioned on that page, are based on the ideal staking rate. That chart is based on an ideal staking rate of 50%, so the numbers will be a bit off (current ideal staking rate on Polkadot is 52.5%).

@laboon, I note your stealth edit substantially revised this from what it was previously.  To my mind that speaks against your assertion that this is obvious or trivial.  

However, you may disagree.  Can you update with any additional feedback or clarify what your remaining objections are?

laboon replied:
I initially used the old algorithm (based on number of auctions, not number of parachain slots) to calculate the ideal staking rate. I realized this rather quickly after posting and updated it, but it looks like you responded even more quickly. That was not some sort of purposeful "stealth edit", but sure, I will admit I made a mistake there.

Here's the correct function in R:

```
# Function to calculate ideal staking rate
calculate_ideal_staking_rate <- function(auctioned_slots) {
  # 30% reserved for up to 60 slots.
  auction_proportion <- min(auctioned_slots, 60) / 200
  
  # Therefore the ideal amount at stake (as a percentage of total issuance) is 75% less the
  # amount that we expect to be taken up with auctions.
  ideal_stake <- 0.75 - auction_proportion
  
  return(ideal_stake)
}
```

This shows the correct ideal staking rate of 0.525.

It is unfortunate that the ideal staking rate is not easily available by querying chain state, I agree with you there.  I don't think that it would be a bad idea to have all of this information more easily visible. But I think this grant is very expensive for what would be produced. Something I would like from a grant like this would be:

1. A much-reduced cost
2. A web interface instead of Excel, or perhaps even a Python script or similar

taqtiqa-mark replied:
Thanks for the clarification,

> Something I would like from a grant like this would be:
>
> * A much-reduced cost

Said every business, every where, every time :)  

The reality is teasing out these insights takes skill and experience.  There is a reason I exposed the 25% shift in 'your' staking rate while it was never disclosed - no one else noticed it.  I knew what to look for, where to look, and what to do with what I was looking for.  Here we have a classical information asymmetry. However, the previous grant, I was told (you can review the thread), was the way to reduce your risk that I was overpriced.  All I can say is, that is the price of these services.

> * A web interface instead of Excel, or perhaps even a Python script or similar

There are two problems with this.  First, this would perpetuate the root problem.  

It seems it isn't self evident, so let me explain, and perhaps add this to the application, if you say it is persuasive.  

One perennial difficulty on software projects is that a developer, through no ones fault or lack of ability, struggles to understand the business need that is addressed by the code they are working on.  Trying to impart that understanding is a major challenge and expense. Multiple developers, business analysts and project managers are put on projects to try and catch the cases where something brilliant for the code base, is dumb for the business.  It's hard to explain further so I give an example below, but this arises frequently in complex businesses (no matter how small the project) or large projects (no matter how simple the business).  

In my experience the only practical bridge between these two worlds (the business and the computer) is an Excel workbook.  The business (SMEs, business analysts, project managers) has to articulate precisely the outputs they expect for given inputs, and in full context.  

This is not trivial to do.  
For a concrete example:  In the case of the Opal project, this system model was a workbook with more than 100 worksheets (it was built as part of the go-live team so wasn't designed to be maintained), each of them 100's of  columns and 1000's of rows (modeling the impact of consumer uptake behavior on the capacity of devices across the network).  That model predicted device saturation 2yrs in advance, led to contract letters being exchanged, and no cost to the client when the vendor had to make urgent/last minute changes.

The reasons not to use the approaches you advocate are the tool has to be usable
- everyday
- in almost all circumstances
 
As you'll notice for the application we can't quiet achieve this decentralized nirvana at the outset.  There are block explorers and data stored outside the workbook. I address that in the future work section.

Second, such an approach would fail to be:
- an audit tool
- accessible to the widest range of users

One use case I should have emphasized: Anyone proposing to change the rewards calculation could be asked to implement the proposal in the 'model'/workbook.  In that use case you'd expect the model to evolve to capture more features.

Let me know if any of that is persuasive.

taqtiqa-mark replied:
@laboon, please see my response to @semuelle above.  I address ths issues there:

> * A much-reduced cost
> 
> * A web interface instead of Excel, or perhaps even a Python script or similar

Would also appreciate a 1-liner response indicating objections maintained, or objections withdrawn. Or some such.

laboon replied:
My objections remain even after your responses to @semuelle .

* I am not aware of any business asking for a tool like this.
* This data can be accessed relatively easily already via Subscan. I put together a quick version this morning in Python, which obviously doesn't cover everything but shows an easy way to get this data which can then be imported into Excel or whatever tool people want to use: https://gist.github.com/laboon/54709b830dffc54ce5f6d5554b82767a 
* Having an alternative implementation of the inflation calculation would be useful, but this is something that I think 
* The chart on the Wiki could be improved. It's based on a screenshot provided by the research team based on the function plotted here ( https://www.desmos.com/calculator/2om7wkewhr ). i_ideal was changed from 0.2 to 0.1 before launch, so the numbers don't line up, but I'll file an issue on the Wiki to fix it.
* The staking rewards calculations will likely change very dramatically in the near future with the new Polkadot changes coming (e.g. coretime), so this tool would require major changes in the very near future.

I appreciate the amount of work you are putting into this grant application but I don't think as it stands it will get an approval from me. Other members of the Grants Committee may disagree.

taqtiqa-mark replied:
> My objections remain even after your responses to @semuelle .
 
Noted.
 
> - I am not aware of any business asking for a tool like this.

I have never suggested there has been or is business interest.  I have always pitched this as an end-user tool. I did digress to point out that it would also be useful in a developer workflow, but emphasised the audience is as wide as possible. 

> - This data can be accessed relatively easily already via Subscan. ...

Yes, I was very clear about the data source [when I identified the problem](https://forum.polkadot.network/t/change-in-stake-rewards-rate-2023-jan-13-and-14/2387/1) this application addresses.  Consider your previous confusion about the data required, taking into account your position, role, and experience. Also consider the lack of response to my request to confirm the data I provided.  With that in mind I don't believe the evidence supports your contention 'this is all trivial'.  Obviously you'll disagree. But you can't point your peers to evidence. 

I do expect you'll try to contend your Python script is proof. And it is. But only if one disregards:

- Your earlier statements (stealth edited)
- The fact this 25% change went unannounced/undetected in the release notes and messaging.
- The fact this 25% change went undetected for 3 months
- The fact my attempts to have my data and calculations validate were ignored for a further 3+ months (no reflection on Kian, I expect he is working as directed)
- It does not address the same audience
- It treats the whole schedule as a single point.

> - Having an alternative implementation of the inflation calculation would be useful, ...

Why? What problem would it solve? What evidence do you have of such a problem? Do you get my point?

> ... but this is something that I think can be provided at a much lower cost.

Of course. Everything is obvious after it has been described.

> - ~~The chart on the Wiki could be improved. It's based on a screenshot provided by the research team based on the function plotted here ( https://www.desmos.com/calculator/2om7wkewhr ). i_ideal was changed from 0.2 to 0.1 before launch, so the numbers don't line up, but I'll file an issue on the Wiki to fix it.~~ EDIT: Disregard - the numbers were already fixed on the chart

**This is beyond objectionable conduct. It is outrageous.**

If this was [a RFP](https://grants.web3.foundation/docs/RFPs/Open/uptane-for-substrate-design-and-scope) it would not be objectionable. But in the course of a live grant it is plagiarism pure and simple. Ask any serious research student or supervisor, and they will likely tell you the key to an important result is identifying an important problem.  In my experience to steal someones insights/ideas or background work that went into identifying a problem is plagiarism.  

We'll now discover if the W3F anti-plagiarism policy has substance.  Particularly if the responsible actor is @laboon, in light of the previous stealth editing.

> - The staking rewards calculations will likely change very dramatically in the near future with the new Polkadot changes coming (e.g. coretime), so this tool would require major changes in the very near future.

A good faith reviewer would seek the scope be adjusted to account for this event.  However, not all treasury proposals go forward. If the upcoming changes are as radical as you suggest they may be rejected. I'd need to think about how to express such a contingent deliverable.

> I appreciate the amount of work you are putting into this grant application but I don't think as it stands it will get an approval from me. 

Yes, I had reached that conclusion too.

> Other members of the Grants Committee may disagree.

It will be interesting/informative to see how this unfolds.

takahser replied:
@taqtiqa-mark Regarding the inflation model graph you [mentioned](https://github.com/w3f/Grants-Program/pull/1943#discussion_r1315590207), the axis descriptions are included in the text of the [wiki page](https://github.com/w3f/polkadot-wiki/blob/master/docs/learn/learn-inflation.md#inflation-model). However, I agree the axes could be labeled directly on the chart as well.
  ![image](https://github.com/w3f/Grants-Program/assets/5393704/5d3fd515-d4c6-4255-83b7-dcb4dd8e7bfc)

laboon commented on applications/decentralized-pallet-staking-audit-tool.md:
```diff
@@ -0,0 +1,374 @@
+
+### Overview
+
+A decentralized tool (Excel/OpenOffice workbook) to allow developers, journalists/bloggers, business analysts, regulators and users to confirm the proper functioning of a program of
```
This information seems already available pretty easily for me -  for user accounts, they can use the  [Staking Rewards Collector](https://github.com/w3f/staking-rewards-collector)  which makes a CSV file.  Or systemically, you can get the data from [Subscan](https://polkadot.subscan.io/event?address=&module=staking&event=erapaid&startDate=&endDate=&startBlock=&endBlock=&timeType=date&version=9430). At this point it's a pretty simple calculation to calculate the rest of the parameters.

laboon commented on applications/decentralized-pallet-staking-audit-tool.md:
Why do you need to send transactions or run a Substrate node for this? If you are relaying on Subscan (which it looks like you are elsewhere), you are already dependent on their nodes.

taqtiqa-mark commented:
> But I have one initial question: Why was the price suddenly increased from 16k to 75k?

The scope is different.

> I also find the estimated workload to develop an Excel/OpenOffice workbook, as well as the daily rate, very high to be honest.

Fair enough. Can you reveal the budget you consider reasonable for this?

taqtiqa-mark commented:
> I also find the estimated workload to develop an Excel/OpenOffice workbook, as well as the daily rate, very high to be honest.

I'll push back here and point to a couple of pieces of evidence that suggest you are underestimating the care that needs to be taken (this is in addition to the evidence provided in the application):

- @laboon makes an assertion that some data is trivially inferred. 
- Then he has to edits this claim, showing he thought the current ideal staking rate was 45%, 
- now thinks it is 52.5%. 
- In both cases he is eyeballing off an imaginary chart.  

Not ideal for a system that invites the unsophisticated to commit in the order of USD 600-700 million.

I was the data lead on the Opal project rolling out a smart card on, rail, bus, ferry and light rail, across the whole of NSW, Australia.  Throughout that roll out the system was in development. Nonetheless, there was only myself and one other analyst to support the design team, the internal reporting, paliamentary reporting, privacy commissioner, and auditor general. As well as all the transport operators.

I don't believe I have mis-estimated the work required.  It is possible you may not understand what is required to build a model of a system such as this - that would also explain the current state of affairs.

My applications have been a good faith attempt to help remedy some serious defects and deficiencies.  

It is getting a little tiresome to be constantly met with push-back that is either wholly ignorant of basic information (Jonas3w on the state of token economic research) or reveals in the comment the precise problem that this project seeks to address (in the case of laboon in this thread)

taqtiqa-mark commented:
> > I also find the estimated workload to develop an Excel/OpenOffice workbook, as well as the daily rate, very high to be honest.
> 
> Fair enough. Can you reveal the budget you consider reasonable for this?

@Noc2, perhaps you could point to one or more projects of similar difficulty and applicant of similar skill and experience, where you found the budget acceptable?

semuelle reviewed (commented):
Thank you for the application, @taqtiqa-mark. I appreciate the level of detail you put into your analyses and argumentation. I tend to agree with my colleagues, though, that this price tag is extraordinary. We occasionally get applications with monthly rates of 12,000-13,000 USD. You are asking for almost double that.

I'm also concerned about the use of spreadsheets. Can you version control all aspects of it, i.e. the spreadsheet, scripts and tests? Is this data structured in a way that new users could pick changes apart? Please correct me if I'm wrong, but an open source data analysis/scripting tool like R's Notebook/Markdown, Shiny or Jupyter might work better in this regard. It could also be hosted on a server for those who don't mind, which would help with verifying results when, inevitably, there will be iterations and multiple versions of it floating around.

taqtiqa-mark commented:
> Thank you for the application, @taqtiqa-mark.

No worries.

> I tend to agree with my colleagues, though, that this price tag is extraordinary.

Sure, as I asked @Noc, please point to applications you consider to be comparable and applicants that have comparable skill and expereince.

To save an iteration:  I believe the comparable project would need to address an issue that if unaddressed prosed a threat to project similar to the issue this application addresses.
Specifically, 

1. Investor rewards were altered by 25% with no announcement or warning.  There is no reason to believe further such changes aren't possible.  If users loose confidence in this aspect of the project they will seek alternatives.  This tool proposes a tool where the most naive user can use on almost any desktop without installing additional software (Python, Jupyter, etc.) and confirm the rewards/dilution they are engaging with.
2.  While this proposal is and will remain pitched at the LCD end user, it could (should) lead to a radical change in the development work flow of a central element of any relay chain (since it will reside in the pallet-staking crate) - no change should be made to the reward/dilution code without first being made to the audit workbook.  Once such a change is made, and socialized from there, it can be implemented and the implementation validated against the audit tool results.

> I'm also concerned about the use of spreadsheets. Can you version control all aspects of it, i.e. the spreadsheet, scripts and tests? 

I could store the XML version for the workbook in the repository.  Workbook tests are mocked in the proposal. The script tests are addressed in the application (bats). 

> Is this data structured in a way that new users could pick changes apart?

As a workbook they could, ideally pick every calculation apart.  The data storage is an issue - mainly because there aren't (currently) clean ways to ingest blockchain data (see future/extensiosn section).  Ideally, a chain would provide an audit record containing all the onchain data relevant to validating/auditing the chain operation.  But this is a chicken and egg problem - there is no widely distributed audit tool that needs and audit record.  If this tool proves propular maybe a chain will step forward to meet the need.  Who knows - right now it is proving very difficult to persuade people one would expect to need the least persuaion :)

> Please correct me if I'm wrong, but an open source data analysis/scripting tool like R's Notebook/Markdown, Shiny or Jupyter might work better in this regard. 

> It could also be hosted on a server for those who don't mind, ...

This is the usual argument for centralization - developer convenience.  IIUC, it buries the problem.  Personally I would have my chain audit tool front and centre.  **_Changes to their rewards/dilution exposure are a critical feature to alert users about_**, not a nuisance to hide behind a server.

> ... which would help with verifying results when, inevitably, there will be iterations and multiple versions of it floating around.

As above these are critical details to be exposed and discussed.  If your rewards/dilution changes so frequently that it challenges an end users ability to understand and keep up with then you really need to address that, rather then hide those changes from them behind a centralized server they will likely never visit.

taqtiqa-mark commented:
> I will mark it as ready for review and share it with the rest of the team.

@noc2 could you clarify one thing for me.  Does approval require reviewer unanimity or just a majority approval?

In the case of unanimity, addressing reviewer B's objections becomes moot if you haven't persuaded Reviewer A.

semuelle commented:
> Does approval require reviewer unanimity or just a majority approval?

Neither. Five committee members must approve a Level 3 application, regardless of what other committee members think.

taqtiqa-mark commented:
> > Does approval require reviewer unanimity or just a majority approval?
> 
> Neither. Five committee members must approve a Level 3 application, regardless of what other committee members think.

Hmmm, okay now I'm confused by the neither statement. So there are committee members and reviewers?  What is the relation between them, and who is providing feedback here?

laboon commented:
> > > Does approval require reviewer unanimity or just a majority approval?
> > 
> > 
> > Neither. Five committee members must approve a Level 3 application, regardless of what other committee members think.
> 
> Hmmm, okay now I'm confused by the neither statement. So there are committee members and reviewers? What is the relation between them, and who is providing feedback here?

By "neither", he means that neither of the cases you mentioned are correct - neither unanimity nor majority approval is required. Rather, a certain number of committee members must approve a grant, based on the level of the grant.

Anyone can "review" a grant, so the term "reviewer" you used is ambiguous, but correcting it earlier probably seemed a bit pedantic and people assumed you meant "Grants Committee member". But to clarify, the term "reviewer" is not terminology in the W3F Grants Program used except in the informal sense of "one reviewing an application".

You can see who is on the Grants Committee in the README of this repo: https://github.com/w3f/grants-Program/#w3f-grants-committee

Web3 Foundation also has Grant Evaluators who have been hired to focus specifically on evaluating grants. You can see a list of them here: https://github.com/w3f/grants-Program/#w3f-grants-evaluators External evaluations can also be done, but these must be reviewed by W3F Grant Evaluators.

Currently, all Grant Evaluators are also on the Grants Committee, but this is not necessarily the case. See the links above for more details on how it works.

takahser reviewed (changes requested):
@taqtiqa-mark thanks for your interest in our grants program. However, I have some concerns about the proposal:

- An Excel implementation would not meet the open source requirements of the Grants Program. Here's a quote from the [guidelines](https://github.com/w3f/Grants-Program/tree/42b0b57d82e7f780c0dea0c093e99e17952a9513#guidelines):
  > All code produced as part of a grant must be open-sourced, and it must also not rely on closed-source software for full functionality.

- An OpenOffice implementation is also not ideal since OpenOffice is a discontinued project.
- I would recommend exploring a web interface tailored to this use case rather than a spreadsheet program. Binary workbook files make version control and decentralization more challenging compared to source code. Also, a web UI has the potential to be much more user-friendly and accessible, also w.r.t. device compatibility.
- A web app hosted on IPFS could provide many of the desired decentralization benefits while being easier to version and update.
- The requested budget of $75k USD seems excessive compared to grants of similar scope. For reference of what other teams charged in the past, I would encourage reviewing the list of previously funded projects and rates [here](https://github.com/w3f/Grants-Program/blob/master/applications/index.md).
- Finally, there are some unresolved concerns related to the necessity and sustainability of a tool like this - Existing solutions can already provide the required data, and major upcoming changes to the network also risk quickly rendering a new tool outdated.

For these reasons, I'm going to abstain from approving this proposal.

semuelle commented:
Hi again, @taqtiqa-mark. Thanks again for your application. After careful deliberation, the committee has decided not to support your application at this point. I want to reiterate that everyone appreciates the time and effort you put into your research and application. I hope you'll still be able to make this project a reality. In any case, best of luck and feel free to apply again in the future.
